libs/ml/include/ml/ops/squeeze.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/ops/ops.hpp"
+
+#include <cassert>
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <class T>
+class Squeeze : public fetch::ml::ops::Ops<T>
+{
+public:
+  using TensorType    = T;
+  using DataType      = typename TensorType::Type;
+  using SizeType      = typename TensorType::SizeType;
+  using VecTensorType = typename Ops<T>::VecTensorType;
+  using SPType        = OpSqueezeSaveableParams<T>;
+  using MyType        = Squeeze<TensorType>;
+
+  Squeeze() = default;
+
+  explicit Squeeze(SPType const &sp)
+    : Ops<T>(sp)
+  {}
+
+  ~Squeeze() override = default;
+
+  std::shared_ptr<OpsSaveableParams> GetOpSaveableParams() override
+  {
+    auto sp = std::make_shared<SPType>();
+    return sp;
+  }
+
+  std::shared_ptr<fetch::ml::ops::Ops<TensorType>> MakeSharedCopy(
+      std::shared_ptr<fetch::ml::ops::Ops<TensorType>> me) override
+  {
+    FETCH_UNUSED(me);
+    assert(me.get() == this);
+
+    auto copyshare = std::make_shared<MyType>(*this);  // calls default copy constructor of MyType
+
+    return copyshare;
+  }
+  /**
+   * Squeeze removes dimensions with size 1
```

Comment:
this is a vague and misleading description.
Squeeze removes the first size 1 dimension encountered starting at the trailing edge. If no dimensions are size 1, it will throw.
It might be worth changing this latter behaviour in the op, relative to the maths. In the maths maybe we want to throw if we can't squeeze, whereas in the ML we might want to do nothing if we can't squeeze. For now we don't have to change behaviour, but this comment could use a little more detail